src/main/ts/core/Changelog.ts

```diff
@@ -0,0 +1,361 @@
+const parseRelease = (source: TextLines, release: Heading, first: boolean): E.Either<string[], Release> => {
+  let meta: ReleaseMeta | undefined;
+  const errors: string[] = [];
+  const headerText = extractText(source, release.header);
+  const header = blockRange(source, release.header);
+  let m: RegExpMatchArray | null = null;
+  if ((m = unreleasedRe.exec(headerText)) && (m.groups?.lbkt?.length ?? 0) === (m.groups?.rbkt?.length ?? 0)) {
+    if (!first) {
+      errors.push('Unexpected "Unreleased" header' + pos(release.header));
+    }
+  } else if ((m = releaseRe.exec(headerText)) && (m.groups?.lbkt?.length ?? 0) === (m.groups?.rbkt?.length ?? 0)) {
+    const versionStr = m.groups?.version ?? '';
+    const dateStr = m.groups?.date ?? '';
+    const versionE = parseVersionE(versionStr);
+    const date = DateTime.fromISO(dateStr, { zone: 'utc' });
+    if (versionE._tag === 'Left') {
+      errors.push('Bad version in header "' + versionE.left + '"' + pos(release.header));
+    } else {
+      meta = { version: versionE.right, date };
+    }
+  } else {
+    errors.push('Unexpected header text to be ' + (first ? '"Unreleased" or ' : '') + '"<version> - <date>"' + pos(release.header));
+  }
+  if (release.preamble.length > 0) {
+    errors.push('Unexpected content under release header' + pos(release.preamble[0]));
+  }
+  return pipe(
+    parseFragment(source, release.subheadings),
+    E.fold(
+      (e) => E.left(errors.concat(e)),
+      (v) => errors.length > 0 ? E.left(errors) : E.right(v)
+    ),
+    E.map((fragment) => {
+      const start = header.start;
+      const sectionCount = fragment.sections.length;
+      const lastSection = sectionCount > 0 ? fragment.sections[sectionCount - 1] : 'Added';
+      const end = fragment[lastSection]?.end ?? header.end;
+      return { ...fragment, meta, header, start, end };
+    })
+  );
+};
+
+const parseReleases = (source: TextLines, releases: Heading[]): E.Either<string[], Release[]> =>
+  EitherUtils.combine(releases.map((release, i) => parseRelease(source, release, i === 0)));
+
+const parseTop = (source: TextLines, top: Top): E.Either<string[], Changelog> => {
+  const errors: string[] = [];
+  if (top.preamble.length > 0) {
+    errors.push('Unexpected content without heading' + pos(top.preamble[0]));
+  }
+  if (top.subheadings.length === 0) {
+    errors.push('No top level heading');
+    return E.left(errors);
+  }
+  if (top.subheadings.length > 1) {
+    errors.push('Unexpected additional top level headings' + pos(top.subheadings[1].header));
+  }
+  const topHeading = top.subheadings[0];
+  if (topHeading.header.level !== 1) {
+    errors.push('First top-level heading is not a h1' + pos(topHeading.header));
+  }
+  const headingText = extractText(source, topHeading.header);
+  if (!/Change[ -]?log/i.test(headingText)) {
+    errors.push('First top-level heading does not contain "Changelog"' + pos(topHeading.header));
+  }
+  return errors.length > 0 ? E.left(errors) : pipe(
```

Comment:
Can this be broken out into an if statement? Ternaries get hard to read when the expressions involved are so large.